utils/__init__.py

```diff
@@ -0,0 +1,138 @@
+import numpy as np
+from numpy.linalg import norm
+from numba import njit
+
+
+class AndersonAcceleration:
+    """Abstraction of Anderson Acceleration.
+
+    Extrapolate the asymptotic VAR ``w`` and ``Xw``
+    based on ``K`` previous iterations.
+
+    Parameters
+    ----------
+    K : int
+        Number of previous iterates to consider for extrapolation.
+    """
+
+    def __init__(self, K):
+        self.K, self.current_iter = K, 0
+        self.arr_w_, self.arr_Xw_ = None, None
+
+    def extrapolate(self, w, Xw):
+        """Return w, Xw, and a bool indicating whether they were extrapolated."""
+        if self.arr_w_ is None or self.arr_Xw_ is None:
+            self.arr_w_ = np.zeros((w.shape[0], self.K+1))
+            self.arr_Xw_ = np.zeros((Xw.shape[0], self.K+1))
+
+        if self.current_iter <= self.K:
+            self.arr_w_[:, self.current_iter] = w
+            self.arr_Xw_[:, self.current_iter] = Xw
+            self.current_iter += 1
+            return w, Xw, False
+
+        U = np.diff(self.arr_w_, axis=1)  # compute residuals
+
+        # compute extrapolation coefs
+        try:
+            inv_UTU_ones = np.linalg.solve(U.T @ U, np.ones(self.K))
+        except np.linalg.LinAlgError:
+            return w, Xw, False
+        finally:
+            self.current_iter = 0
+
+        # extrapolate
+        C = inv_UTU_ones / np.sum(inv_UTU_ones)
+        # floating point errors may cause w and Xw to disagree
+        return self.arr_w_[:, 1:] @ C, self.arr_Xw_[:, 1:] @ C, True
+
+
+@njit
+def spectral_norm(X_data, X_indptr, X_indices, n_samples,
+                  max_iter=100, tol=1e-6):
+    """Compute the spectral norm of sparse matrix ``X`` with power method.
```

Comment:
I'd put all the functions related to operations on sparse operations in a module `sparse_ops.py` and put `AndersonAcceleration` class in its own module `anderson.py`. This way we have a clear separation of all these helper functions.